src/strategy/filesystem.rs

```diff
@@ -0,0 +1,180 @@
+//! Strategy for local filesystem-based updates.
+
+use anyhow::{Context, Error, Result};
+use fn_error_context::context;
+use futures::future;
+use futures::prelude::*;
+use log::trace;
+use serde::{Deserialize, Serialize};
+use std::os::unix::fs::{MetadataExt, PermissionsExt};
+use std::pin::Pin;
+use std::time::{SystemTime, UNIX_EPOCH};
+use tokio::fs as tokio_fs;
+
+/// Struct to parse finalization marker file's JSON content into.
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct FinalizationMarker {
+    /// Unix timestamp of expiry time.
+    not_after: Option<u64>,
+}
+
+/// Strategy for immediate updates.
+#[derive(Clone, Debug, Default, Serialize)]
+pub(crate) struct StrategyFilesystem {}
+
+impl StrategyFilesystem {
+    /// Strategy label/name.
+    pub const LABEL: &'static str = "filesystem";
+    /// Local filesystem path to finalization marker file.
+    pub const FINALIZATION_MARKER_FILE_PATH: &'static str = "/var/lib/zincati/allowfinalize.json";
+
+    /// Check if finalization is allowed.
+    pub(crate) fn can_finalize(&self) -> Pin<Box<dyn Future<Output = Result<bool, Error>>>> {
+        Box::pin(Self::filesystem_allow_finalization(
+            Self::FINALIZATION_MARKER_FILE_PATH,
+        ))
+    }
+
+    /// Try to report steady state.
+    pub(crate) fn report_steady(&self) -> Pin<Box<dyn Future<Output = Result<bool, Error>>>> {
+        trace!("filesystem strategy, report steady: {}", true);
+
+        let res = future::ok(true);
+        Box::pin(res)
+    }
+
+    /// Asynchronous helper function that returns a future indicating
+    /// whether finalization is allowed.
+    async fn filesystem_allow_finalization(finalization_marker_path: &str) -> Result<bool> {
+        if !verify_file_metadata(finalization_marker_path).await? {
+            return Ok(false);
+        }
+
+        if is_expired(finalization_marker_path).await? {
+            return Ok(false);
+        }
+
+        Ok(true)
+    }
+}
+
+/// Verify that finalization marker file exists, is a regular file,
+/// has the correct permissions, and is owned by the root user.
+#[context("failed to verify finalization marker file metadata")]
+async fn verify_file_metadata(path: &str) -> Result<bool> {
+    let attr = tokio_fs::metadata(path).await;
+    // If `path` doesn't exist, return false early.
+    if attr.is_err() {
+        return Ok(false);
+    }
+    let attr = attr.unwrap();
```

Comment:
You can hoist this into the err-check above, using a `match` to avoid the unwrapping.